File: duna-core/src/main/java/io/duna/core/concurrent/future/AbstractFuture.java
package io.duna.core.concurrent.future;

import java.util.concurrent.atomic.AtomicReference;

public abstract class AbstractFuture<T> implements Future<T> {

    private final AtomicReference<FutureState> state;

    public AbstractFuture() {
        this.state = new AtomicReference<>(FutureState.NEW);
    }

    @Override
    public boolean done() {
        return state.get() != FutureState.NEW;
    }

    @Override
    public boolean completed() {
        return state.get() == FutureState.COMPLETED;
    }

    @Override
    public boolean failed() {
        return state.get() == FutureState.FAILED;
    }

    @Override
    public boolean cancelled() {
        return state.get() == FutureState.CANCELLED;
    }

    @Override
    public void cancel() {
        if (!state.compareAndSet(FutureState.NEW, FutureState.CANCELLED))
            throw new IllegalStateException("Future already closed.");
    }

    @Override
    public void complete(T result) {
        if (!state.compareAndSet(FutureState.NEW, FutureState.COMPLETED))
            throw new IllegalStateException("Future already closed.");
    }

    @Override
    public void fail(Throwable error) {
        if (!state.compareAndSet(FutureState.NEW, FutureState.FAILED))
            throw new IllegalStateException("Future already closed.");
    }
}

File: duna-core/src/main/kotlin/io/duna/core/concurrent/AbstractFuture.kt
package io.duna.core.concurrent

import io.duna.core.concurrent.FutureState.*
import java.util.concurrent.atomic.AtomicReference

internal abstract class AbstractFuture<T> : Future<T> {

  private val state = AtomicReference<FutureState>(NEW)

  override fun done() = state.get() != NEW

  override fun completed() = state.get() == COMPLETED

  override fun failed() = state.get() == FAILED

  override fun cancelled() = state.get() == CANCELLED

  override fun cancel() {
    if (!state.compareAndSet(NEW, CANCELLED))
      throw IllegalStateException()
  }

  override fun complete(result: T?) {
    if (!state.compareAndSet(NEW, COMPLETED))
      throw IllegalStateException()
  }

  override fun fail(error: Throwable) {
    if (!state.compareAndSet(NEW, FAILED))
      throw IllegalStateException()
  }
}
